File: virturian_d1.R
library(FSelector)
library(randomForest)

#Load data

setwd("C:\\Users\\gustavo.v.machado\\Downloads")

waveDataRaw = read.csv("waveform.data",header=FALSE,sep=",")

#Prepare: scale features and rename classification variable

waveFeatures = scale(waveDataRaw[,1:21])
waveLabel = waveDataRaw[,22]

waveDataRaw = as.data.frame(cbind(waveFeatures,LABEL=waveLabel))
waveDataRaw$LABEL = as.factor(waveDataRaw$LABEL)

#Separate training set from test set

set.seed(596)
trainIndex = sample(seq_len(nrow(waveDataRaw)), size = floor(0.7 * nrow(waveDataRaw)))

waveTrainSet = waveDataRaw[trainIndex, ]
waveTestSet = waveDataRaw[-trainIndex, ]

#Feature selection with CFS

topFeatures = cfs(LABEL ~.,waveTrainSet)

waveTrainSelFeat = waveTrainSet[, topFeatures]
waveTestSelFeat = waveTestSet[, topFeatures]

#Generate classification model based on random forest

waveTrain = as.data.frame(cbind(waveTrainSelFeat,LABEL=waveTrainSet$LABEL))
waveTrain$LABEL = as.factor(waveTrain$LABEL)

model = randomForest(LABEL ~ .,data=waveTrain,ntree=100,proximity=T)

result = predict(model, waveTestSelFeat)

correctClassificationRate = sum(diag(table(result, waveTestSet$LABEL)))/nrow(waveTestSet)
correctClassificationRate

File: virturian_d1.py
import pandas, numpy, statistics
from sklearn import metrics, svm, feature_selection, cross_validation


def main():
    considerNoise = True
    enableFS = False

    waveDataRaw = loadWaveData(considerNoise)

    f1_scoreList = []

    for x in range(10):

        waveTestSet, waveTrainSet = generateTrainAndTestSets(waveDataRaw)

        model = createClassifier()

        featuresName = selectFeatures(waveTrainSet, model, enableFS)

        model = trainClassifier(waveTrainSet, featuresName, model)

        result, f1_score = classifyAndEvaluate(model, waveTestSet, featuresName)

        f1_scoreList.append(f1_score)

    print("mean: ", statistics.mean(f1_scoreList))
    print("sda: ", statistics.stdev(f1_scoreList))


def createClassifier():

    model = svm.SVC(kernel="linear")

    return model

def generateTrainAndTestSets(waveDataRaw):
    #Generate random trainig and test sets with 70% and 30% of data

    trainIndex = numpy.random.rand(len(waveDataRaw)) < 0.7
    waveTrainSet = waveDataRaw[trainIndex]
    waveTestSet = waveDataRaw[~trainIndex]
    return waveTestSet, waveTrainSet


def selectFeatures(waveTrainSet, model, enableFS):
    setLen = len(waveTrainSet.columns)

    featuresName = list(waveTrainSet.columns.values)
    featuresName = featuresName[:setLen - 2]

    if enableFS:

        #Recursive feature selection

        recursiveFS = feature_selection.RFECV(estimator=model, step=1, cv=cross_validation.StratifiedKFold(waveTrainSet.ix[:, setLen-1], 2),
                      scoring='f1_weighted')
        recursiveFS.fit(waveTrainSet[featuresName], waveTrainSet.ix[:, setLen-1])

        #Get all features names ranked as 1

        featuresName = list(map(lambda x: x[1],filter(lambda x: x[0]==1,zip(recursiveFS.ranking_, featuresName))))

        print("Selected features:")
        print(featuresName)

    return featuresName

def classifyAndEvaluate(model, waveTestSet, featuresName):
    setLen = len(waveTestSet.columns)

    result = list(model.predict(waveTestSet[featuresName]))

    print("Result:")
    print(metrics.classification_report(waveTestSet.ix[:, setLen - 1], result))

    return result, metrics.f1_score(waveTestSet.ix[:, setLen - 1], result)


def trainClassifier(waveTrainSet, featuresName, model):
    setLen = len(waveTrainSet.columns)

    model.fit(waveTrainSet[featuresName], waveTrainSet.ix[:, setLen-1])

    return model


def loadWaveData(considerNoise):

    file = "waveform-+noise.data" if considerNoise else "waveform.data"

    waveFileName = "C:\\Users\\gustavo.v.machado\\Downloads\\" + file
    waveDataRaw = pandas.read_csv(waveFileName, sep=',', header=None, names=None)


    return waveDataRaw


if __name__ == "__main__":
    main()
